closed/make/Javadoc.gmk

```diff
@@ -0,0 +1,490 @@
+# When pages are being generated openj9 extensions (under the docs/platform
+# directory), the correct footer is specified directly in the javadoc command.
+#
+# The OpenJ9 classes are built from the j9jcl_sources and closed/src
+# directories.
+# The packages are beneath directories called /share/classes (or /unix/classes,
+# /windows/classes etc.) which do not appear in the javadoc tree structure.
+# The commands below replace the footer in the javadoc html pages for the
+# following files:
+# 1. The openj9 java source files which override the openjdk versions
+# 2. The openj9.xxxx modules, including the javadoc generated files such as
+#    enum class pages, module-summary.html, package-summary.html etc.)
+# A temporary script file is used to avoid the xargs command line length limit
+# of 255 chars.
+
+OPENJ9_SUBST_BOTTOM = $(subst &,\&,$(OPENJ9_JAVADOC_BOTTOM))
+
+#############################################################################
+# sed command to replace the footer in the jdk management javadoc files.
+MGMT_BOTTOM_REGEX = $(subst .,\.,$(MGMT_BOTTOM))
+OPENJ9_MGMT_SUBST_COMMAND = $(SED) -e '\''s|$(MGMT_BOTTOM_REGEX)|$(OPENJ9_SUBST_BOTTOM)|g'\''
+
+#############################################################################
+# sed command to replace the footer in the core api javadoc files.
+# Openjdk Core API doc is generated with the -Xdocrootparent javadoc option,
+# so apply the same substition to the page footer before matching.
+COREAPI_BOTTOM_REGEX = $(subst {@docroot}/,$(subst .,\.,$(DOCS_BASE_URL)),$(subst .,\.,$(COREAPI_BOTTOM)))
+OPENJ9_COREAPI_SUBST_COMMAND = $(SED) -e '\''s|$(COREAPI_BOTTOM_REGEX)|$(OPENJ9_SUBST_BOTTOM)|g'\''
+
+################################################################################
+OPENJ9_JAVADOC_MARKER := $(TEMPDIR)/_javadoc_openj9.marker
+OPENJ9_JAVADOC_SCRIPT := $(TEMPDIR)/_javadoc_openj9_footer_replace.sh
+OPENJ9_JAVADOC_TEMP  := $(TEMPDIR)/_javadoc_temp.html
+
+################################################################################
+# Use a marker file as target with dependencies which ensure footers are not
+# updated until the pages have been updated by the openjdk javadoc targets. 
+
+OPENJ9_JAVADOC_SCRIPT_BODY = \
+    for path in "$$@" ; do \
+      if [ -f "$(COREAPI_DOCSDIR)/$$path" ] ; then \
+        $(CP) "$(COREAPI_DOCSDIR)/$$path" "$(OPENJ9_JAVADOC_TEMP)" ; \
+        $(OPENJ9_COREAPI_SUBST_COMMAND) < "$(OPENJ9_JAVADOC_TEMP)" > "$(DOCS_OUTPUTDIR)/$$path" ; \
+        $(RM) "$(OPENJ9_JAVADOC_TEMP)" ; \
+      fi ; \
+      if [ -f "$(JRE_API_DOCSDIR)/management/extension/$$path" ] ; then \
+        $(CP) "$(JRE_API_DOCSDIR)/management/extension/$$path" "$(OPENJ9_JAVADOC_TEMP)" ; \
+        $(OPENJ9_MGMT_SUBST_COMMAND) < "$(OPENJ9_JAVADOC_TEMP)" > "$(JRE_API_DOCSDIR)/management/extension/$$path" ; \
+        $(RM) "$(OPENJ9_JAVADOC_TEMP)" ; \
+      fi ; \
+    done \
+    #
+
+$(OPENJ9_JAVADOC_MARKER) : $(COREAPI_DOCSDIR)/index.html openj9-mgmtdocs
```

Comment:
`openj9-mgmtdocs` appears to be a phony target (which should explicitly be declared). Using a phony target as a prerequisite will force execution of this rule unconditionally.

Reply:
openj9-mgmtdocs is declared as PHONY in make/Javadoc.gmk as a result of it being included in ALL_OTHER_TARGETS.
It was done this way simply because that is the model used by the openjdk build for the various doc areas - a phony target with a dependency on the index.html file for that area.

Reply:
I'm fine if that target is declared PHONY elsewhere, but the unconditional execution of this rule is a problem that needs to be fixed.

Reply:
My assumption is that architected way for additional javadoc targets to be added to the build is to add them to the list of targets in the ALL_OTHER_TARGETS macro, which are dependencies of the 'otherdocs' target which is a dependency of the 'docs' target.  The ALL_OTHER_TARGETS  targets are all declared as phony in the including make/Javadoc.gmk, so they are all phony dependencies.
Or have I misunderstood your point?

Reply:
It's fine for PHONY targets to depend on real files (as long as the script for those phony targets does nothing substantial), but not the other way around.

Reply:
OK.  So my issue is how to use the ALL_OTHER_TARGETS as a means of adding the extra javadoc targets to the build, but to be sure of only executing the target to replace the footers when the core api javadoc and the management/extensions javadoc generation has completed.
The code currently does that by depending on openj9_mgmtdocs, which depends on mgmtdocs (in the openjdk make/Javadoc.gmk file) which depends on coredocs (in the openjdk make/Javadoc.gmk file).

Reply:
I've replaced the phony target dependencies with the file targets which those phony targets depended on.  @keithc-ca - please can you take another look.